### PR TITLE
Allow duplicate items to be awarded

### PR DESCRIPTION
**Pull request by dukeofphilberg**
This PR proposes to color an item icon green in the auction window when it has been awarded. This behaviour is similar to the bidding window where the item is also colored green when a bid has been placed on it.

In-game image:
image

### CODE REVIEW

**Comment by lantisnt** [comment]
What would be the value added here as by default items are intended to be removed from window?
What if we would want to award multiple items (so the reason to keep info)? Then marking in green missed the point.
Happy to hear your input

**Comment by dukeofphilberg** [comment]
> What would be the value added here as by default items are intended to be removed from window? What if we would want to award multiple items (so the reason to keep info)? Then marking in green missed the point. Happy to hear your input

The idea behind it is if you don't check the "remove on award" then you still know which items you distributed. This is something we need to do when distributing duplicates of an item (e.g. tier tokens.) I do see the problem when you have duplicates of the same item it doesn't quite makes sense to already color it green.

**Comment by lantisnt** [comment]
What if we combined this idea with a display of amount to award?
Green when the count goes to 0, or maybe a checkmark? A count displayed over otherwise

**Comment by dukeofphilberg** [comment]
> What if we combined this idea with a display of amount to award? Green when the count goes to 0, or maybe a checkmark? A count displayed over otherwise

That would be a good solution. It's already possible to award the same item multiple times so that's set on the backend side too. It'll require the `AuctionItem` object to have a list of entries though, rather than now just the one. But that's easily solvable by holding a list.

Could we create a text just below the bidding window "Awarded x out of n [item]"?

image

**Comment by lantisnt** [comment]
That doesn't need to be a list as we dont track the actual Item objects (as discerned by ItemGUID) and we dont really need to.
A simple Item instance counter (auctionItem created with count=1) would suffice. On adding existing Item to auction, when we discard it, we would increment the count instead and decrement on award. A nice option to add would be a warning if we want to award more than we have added (but not blocking the award completely).

When it comes to display i'm hesitant where to put it. Adding additional row seems excessive. Maybe a value next to award on the button that's updated accordingly?

``Award (0/1)`

**Comment by dukeofphilberg** [comment]
> That doesn't need to be a list as we dont track the actual Item objects (as discerned by ItemGUID) and we dont really need to. A simple Item instance counter (auctionItem created with count=1) would suffice. On adding existing Item to auction, when we discard it, we would increment the count instead and decrement on award. A nice option to add would be a warning if we want to award more than we have added (but not blocking the award completely).
> 
> When it comes to display i'm hesitant where to put it. Adding additional row seems excessive. Maybe a value next to award on the button that's updated accordingly?
> 
> `Award (0/1)`

Sounds like a plan. I'll try and implement accordingly. Maybe we'll need another way to decrement the `count` seperate from awarding (for whatever reason). Maybe rightclicking on the item as it removes the item now. Doing so we could also decrease the count, and only remove the item when `count = 0`.

**Comment by dukeofphilberg** [comment]
I've done some preliminary testing and the idea seems to work. The text on the button is cut off though. Currently it shows (`x items awarded`/`y items to award`). Upon awarding an item the `x` decreases by one. When `x` is zero and the `Remove on award` is checked the item is properly removed. Will need have some additional logic when the check is not enabled so that we can't keep awarding items when there's none left.

image

File: ClassicLootManager/Modules/Auctioning/AuctionManager/MultiGUI.lua [hunk 1]
```diff
@@ -265,6 +265,14 @@ local function CreateLootList(self)
 end
 
 local function GenerateAwardOptions(self)
+    local btnAwardText = CLM.L["Award"]
+    if self.auctionItem then
+        btnAwardText = string.format(
+            CLM.L["Award (%d/%d)"],
```
Comment (lantisnt) [inline ClassicLootManager/Modules/Auctioning/AuctionManager/MultiGUI.lua:271]:
```suggestion
            CLM.L["Award"] .. " (%d/%d)",
```

File: ClassicLootManager/Modules/Auctioning/Common/AuctionInfo.lua [hunk 2]
```diff
@@ -236,8 +236,9 @@ end
 
 local function AddItemInternal(self, item)
     if self.items[item:GetItemID()] then
-        LOG:Message("%s already in auction list.", item:GetItemLink())
```
Comment (lantisnt) [inline ClassicLootManager/Modules/Auctioning/Common/AuctionInfo.lua:239]: Keep this line but add information that we are adding new:
`LOG:Message("%s already in auction list. Increasing the amount to award.", item:GetItemLink())`

File: ClassicLootManager/Modules/Auctioning/Common/AuctionInfo.lua [hunk 3]
```diff
@@ -47,7 +47,7 @@ end
 
 local function Clear(self)
     for _, item in pairs(self.items) do
-        item:Clear()
+        --item:Clear()
```
Comment (lantisnt) [inline ClassicLootManager/Modules/Auctioning/Common/AuctionInfo.lua:50]: Why do you remove this one?

**Review by lantisnt — CHANGES_REQUESTED** [review]
Looks good in general. A small nitpick: Instead of having
`self.amountToAward` and `self.amountAwarded` I would rather have only 1 variable e.g.
`self.total` that would increment on adding to auction and decrement on award, thus we could simply check in the end if items left is 0.

File: ClassicLootManager/Modules/Auctioning/Common/AuctionInfo.lua (hunk 3 above)
Comment (dukeofphilberg) [reply ClassicLootManager/Modules/Auctioning/Common/AuctionInfo.lua]: Upon starting bidding, all items get Cleared, which means that the `total` also gets reset. I'll deal with this in another way.

**Comment by dukeofphilberg** [comment]
I'm pretty happy with it now. Code is much cleaner keeping just one variable to track the amount to award and fits better within the text of the button 👍

image

The only thing that might be confusing is that the Disenchant button clears the total amount, instead of stepping it down one, like awarding an item does. What do you think?

**Comment by lantisnt** [comment]
It should decrement indeed. Also ctrl click should still remove it totally but new ctrl+shift probably decrement?

**Review by lantisnt — APPROVED** [review]
I'll conditionally merge this one and do some further rework to handle some corner cases as it will be easier for me to do them rather than explain them 😆

**Pull request merged**